File: 106.Construct Binary Tree from Inorder and Postorder Traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} inorder
 * @param {number[]} postorder
 * @return {TreeNode}
 */
var buildTree = function(inorder, postorder) {
    if (inorder.length) {
        const rootValue = postorder[postorder.length - 1];
        const rootIndex = inorder.indexOf(rootValue);
        
        const leftTreeInOrder = inorder.slice(0, rootIndex);
        const rightTreeInOrder = inorder.slice(rootIndex + 1);
        const leftTreePostOrder = postorder.slice(0, leftTreeInOrder.length);
        const rightTreePostOrder = postorder.slice(leftTreeInOrder.length, postorder.length - 1);
        
        const node = new TreeNode(rootValue);
        node.left = buildTree(leftTreeInOrder, leftTreePostOrder);
        node.right = buildTree(rightTreeInOrder, rightTreePostOrder);
        return node;
    }
    return null;
};

File: Golang/Tree/106.construct-binary-tree-from-inorder-and-postorder-traversal.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func buildTree(inorder []int, postorder []int) *TreeNode {
	var indexOf = func(slice []int, target int) int {
		for i, v := range slice {
			if v == target {
				return i
			}
		}
		return -1
	}
	var recursive func(i, p []int) *TreeNode
	recursive = func(i, p []int) *TreeNode {
		length := len(i)
		if length == 0 {
			return nil
		} else if length == 1 {
			return &TreeNode{
				Val:   i[0],
				Left:  nil,
				Right: nil,
			}
		}
		rootVal := p[length-1]
		spliterIndex := indexOf(i, rootVal)
		iLeft := i[:spliterIndex]
		iRight := i[spliterIndex+1:]
		pLeft := p[:len(iLeft)]
		pRight := p[len(iLeft) : len(p)-1]
		return &TreeNode{
			Val:   p[length-1],
			Left:  recursive(iLeft, pLeft),
			Right: recursive(iRight, pRight),
		}
	}
	return recursive(inorder, postorder)
}
